Sources/NIOCore/EventLoopFuture+AssumeIsolated.swift

```diff
@@ -0,0 +1,510 @@
+    func scheduleTask<T>(
+        deadline: NIODeadline,
+        _ task: @escaping () throws -> T
+    ) -> Scheduled<T> {
+        self.wrapped.assertInEventLoop()
+        let unsafeTransfer = UnsafeTransfer(task)
+        return wrapped.scheduleTask(deadline: deadline) {
+            try unsafeTransfer.wrappedValue()
+        }
+    }
+
+    /// Schedule a `task` that is executed by this `EventLoop` after the given amount of time.
+    ///
+    /// - parameters:
+    ///     - task: The synchronous task to run. As with everything that runs on the `EventLoop`, it must not block.
+    /// - returns: A `Scheduled` object which may be used to cancel the task if it has not yet run, or to wait
+    ///            on the completion of the task.
+    ///
+    /// - note: You can only cancel a task before it has started executing.
+    /// - note: The `in` value is clamped to a maximum when running on a Darwin-kernel.
+    @discardableResult
+    @inlinable
+    func scheduleTask<T>(
+        in delay: TimeAmount,
+        _ task: @escaping () throws -> T
+    ) -> Scheduled<T> {
+        self.wrapped.assertInEventLoop()
+        let unsafeTransfer = UnsafeTransfer(task)
+        return wrapped.scheduleTask(in: delay) {
+            try unsafeTransfer.wrappedValue()
+        }
+    }
+
+    /// Schedule a `task` that is executed by this `EventLoop` at the given time.
+    ///
+    /// - Note: The `T` must be `Sendable` since the isolation domains of the event loop future returned from `task` and
+    /// this event loop might differ.
+    ///
+    /// - parameters:
+    ///     - task: The asynchronous task to run. As with everything that runs on the `EventLoop`, it must not block.
+    /// - returns: A `Scheduled` object which may be used to cancel the task if it has not yet run, or to wait
+    ///            on the full execution of the task, including its returned `EventLoopFuture`.
+    ///
+    /// - note: You can only cancel a task before it has started executing.
+    @discardableResult
+    @inlinable
+    func flatScheduleTask<T: Sendable>(
+        deadline: NIODeadline,
+        file: StaticString = #file,
+        line: UInt = #line,
+        _ task: @escaping () throws -> EventLoopFuture<T>
+    ) -> Scheduled<T> {
+        self.wrapped.assertInEventLoop()
+        let unsafeTransfer = UnsafeTransfer(task)
+        return wrapped.flatScheduleTask(deadline: deadline, file: file, line: line) {
+            try unsafeTransfer.wrappedValue()
+        }
+    }
+}
+extension EventLoop {
+    /// Assumes the calling context is isolated to the event loop.
+    @usableFromInline
+    func assumeIsolated() -> AssumeIsolatedEventLoop {
+        AssumeIsolatedEventLoop(self)
+    }
+}
+
+extension EventLoopFuture {
+    /// A struct wrapping an ``EventLoopFuture`` that ensures all calls to any method on this struct
+    /// are coming from the event loop of the future.
+    @usableFromInline
+    struct AssumeIsolated: @unchecked Sendable {
```

Comment:
I'm a bit surprised that none of the methods on here return the caller back an `AssumeIsolated` future? This would make chaining much easier. Also, similar to above, it feels strange that we can move the `AssumeIsolated` future across isolation domains because it's `Sendable`. I think we want to make the assumption when we create it and have it not be `Sendable`.